grove/ising/ising_qaoa.py

```diff
@@ -0,0 +1,407 @@
+
+"""
+Finding the minimum energy for an Ising problem by QAOA.
+"""
+import pyquil.api as api
+from pyquil.paulis import exponential_map, PauliSum, PauliTerm
+from functools import reduce
+import pyquil.quil as pq
+from pyquil.gates import H
+from grove.pyvqe.vqe import VQE
+from scipy.optimize import minimize
+from collections import Counter
+from scipy import optimize
+import numpy as np
+import os
+import sys
+
+CXN = api.SyncConnection()
+
+
+class silence(object):
+    """
+    Class that makes functions silent.
+    """
+
+    def __init__(self, stdout=None, stderr=None):
+        self.devnull = open(os.devnull, 'w')
+        self._stdout = stdout or self.devnull or sys.stdout
+        self._stderr = stderr or self.devnull or sys.stderr
+
+    def __enter__(self):
+        self.old_stdout, self.old_stderr = sys.stdout, sys.stderr
+        self.old_stdout.flush()
+        self.old_stderr.flush()
+        sys.stdout, sys.stderr = self._stdout, self._stderr
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._stdout.flush()
+        self._stderr.flush()
+        sys.stdout = self.old_stdout
+        sys.stderr = self.old_stderr
+        self.devnull.close()
+
+
+class QAOA_ising(object):
```

Comment:
this seems like a duplication of the regular QAOA class. Can we reuse that rather than copying it? Or is there any functionality that I am missing?

Reply:
There are few minor differences with respect to the original QAOA class in order for the current Ising implementation to work. Again, I didn't want to edit the original QAOA class, so I just made a new copy with few tweaks.

Reply:
This could potentially become a significant maintenance burden with duplicated code. What are the minor difference? Can you highlight them? It's hard to see just on the PR what the different functionality is.